std/digest/xxh.d

```diff
@@ -0,0 +1,3180 @@
+{
+    ulong* xacc = cast(ulong*) acc; /* presumed aligned */
+    const ubyte* xsecret = cast(const ubyte*) secret; /* no alignment restriction */
+    assert(((cast(size_t) acc) & (XXH_ACC_ALIGN - 1)) == 0, "((cast(size_t) acc) & (XXH_ACC_ALIGN - 1)) != 0");
+    assert(lane < XXH_ACC_NB, "lane >= XXH_ACC_NB");
+    {
+        const ulong key64 = XXH_readLE64(xsecret + lane * 8);
+        ulong acc64 = xacc[lane];
+        acc64 = XXH_xorshift64(acc64, 47);
+        acc64 ^= key64;
+        acc64 *= XXH_PRIME32_1;
+        xacc[lane] = acc64;
+    }
+}
+
+/*!
+ * @internal
+ * @brief Scrambles the accumulators after a large chunk has been read
+ */
+private void XXH3_scrambleAcc_scalar(void* acc, const(void)* secret) @safe pure nothrow @nogc
+{
+    size_t i;
+    for (i = 0; i < XXH_ACC_NB; i++)
+    {
+        XXH3_scalarScrambleRound(acc, secret, i);
+    }
+}
+
+private void XXH3_initCustomSecret_scalar(void* customSecret, ulong seed64) @trusted pure nothrow @nogc
+{
+    /*
+     * We need a separate pointer for the hack below,
+     * which requires a non-const pointer.
+     * Any decent compiler will optimize this out otherwise.
+     */
+    const ubyte* kSecretPtr = cast(ubyte*) XXH3_kSecret;
+    static assert((XXH_SECRET_DEFAULT_SIZE & 15) == 0, "(XXH_SECRET_DEFAULT_SIZE & 15) != 0");
+
+    /*
+     * Note: in debug mode, this overrides the asm optimization
+     * and Clang will emit MOVK chains again.
+     */
+    //assert(kSecretPtr == XXH3_kSecret);
+
+    {
+        const int nbRounds = XXH_SECRET_DEFAULT_SIZE / 16;
+        int i;
+        for (i = 0; i < nbRounds; i++)
+        {
+            /*
+             * The asm hack causes Clang to assume that kSecretPtr aliases with
+             * customSecret, and on aarch64, this prevented LDP from merging two
+             * loads together for free. Putting the loads together before the stores
+             * properly generates LDP.
+             */
+            ulong lo = XXH_readLE64(kSecretPtr + 16 * i) + seed64;
+            ulong hi = XXH_readLE64(kSecretPtr + 16 * i + 8) - seed64;
+            XXH_writeLE64(cast(ubyte*) customSecret + 16 * i, lo);
+            XXH_writeLE64(cast(ubyte*) customSecret + 16 * i + 8, hi);
+        }
+    }
+}
+
+alias XXH3_f_accumulate_512 = void function(void*, const(void)*, const(void)*) @safe pure nothrow @nogc;
+alias XXH3_f_scrambleAcc = void function(void*, const void*) @safe pure nothrow @nogc;
+alias XXH3_f_initCustomSecret = void function(void*, ulong) @safe pure nothrow @nogc;
+
+immutable XXH3_f_accumulate_512 XXH3_accumulate_512 = &XXH3_accumulate_512_scalar;
+immutable XXH3_f_scrambleAcc XXH3_scrambleAcc = &XXH3_scrambleAcc_scalar;
+immutable XXH3_f_initCustomSecret XXH3_initCustomSecret = &XXH3_initCustomSecret_scalar;
+
+enum XXH_PREFETCH_DIST = 384;
+private void XXH_PREFETCH(const ubyte* ptr) @safe pure nothrow @nogc
+{
+    cast(void)(ptr);
```

Comment:
I'm curious about this. Is that supposed to force prefetching ? I see this pattern everywhere.

Reply:
Yes, it seems so. It's taken from the C sources. This might be optimized away in D, so there is no observable effect. Maybe there is already some intrinisc D function, which is intended for this purpose. The original C code is also heavily augmented with pragmas to allow/disallow inlining of code and get the desired code.

I presume, that we need to rework this code to operate with D as expected. Or throw it away and let it to the optimizer. However, before throwing away code, I should run benchmarks and examine the code ;-)

Reply:
ok, so that's a little red flag. D can prefetch using inline asm and certain vendor (GDC, LDC) can use special intrinsics
for that I suppose.

Reply:
It does only hurt performance, if it is not working right now. After the initial code cleanup, this code must be finetuned to allow prefetching the one or other way.

Reply:
See https://dlang.org/library/core/simd/prefetch.html for a more idiomatic way of prefetching.

Reply:
Is there any kind of D core module, which provides a architecure independant way to trigger prefetch operations? 
Are there any means provided by LDC2, GDC or DMD to inline (or not) a prefetch mnemonic into the code?

I added some TODO and comments to this prefetch code. This issue can be resolved later.